File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-4.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        res = 1
        memo = {}
        d = [(1, 0), (0, 1), (-1, 0), (0, -1)]
        
        def dfs(i, j) -> int:
            if (i, j) in memo:
                return memo[(i, j)]
            maxi = 0
            for x, y in d:
                ii = x + i
                jj = y + j
                if ii >=0 and ii < len(matrix) and jj >=0 and jj < len(matrix[0]) and matrix[ii][jj] > matrix[i][j]:
                    maxi = max(maxi, dfs(ii, jj))
            memo[(i, j)] = maxi + 1
            return maxi + 1
        
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                res = max(res, dfs(i, j))
        print(memo)
        return res

File: Data Structures & Algorithms/decode-ways/submission-4.cpp
class Solution {
public:
    int numDecodings(string s) {
        if (s.size() == 0){
            return 1;
        }
        if(s[0] == '0') return 0;
        int res = numDecodings(s.substr(1));
        if (s.size() >= 2) {
            if ((s[0] == '1') || (s[0] == '2' && s[1] < '7')){
                res += numDecodings(s.substr(2));
            }
        }
        return res;
    }
};
